importer/importer.go

```diff
@@ -369,61 +366,22 @@ func (im *Importer) insertData(conn driver.Conn, bucket, key, db, table string,
 	return err
 }
 
-func (im *Importer) connect() (driver.Conn, error) {
-	var (
-		ctx       = context.Background()
-		conn, err = clickhouse.Open(&clickhouse.Options{
-			Addr: []string{fmt.Sprintf("%s:%s", im.Config.Clickhouse.Host, im.Config.Clickhouse.TCPPort)},
-			Auth: clickhouse.Auth{
-				// Database: "default",
-				Username: im.Config.Clickhouse.Username,
-				Password: im.Config.Clickhouse.Password,
-			},
-			Debug:           false,
-			MaxOpenConns:    im.Config.Insert.MaxOpenConns,
-			MaxIdleConns:    im.Config.Insert.MaxIdleConns,
-			ConnMaxLifetime: time.Second * time.Duration(im.Config.Insert.ConnMaxLifetimeSecs),
-			// ClientInfo: clickhouse.ClientInfo{
-			// 	Products: []struct {
-			// 		Name    string
-			// 		Version string
-			// 	}{
-			// 		{Name: "scratchdb", Version: "1"},
-			// 	},
-			// },
-
-			// Debugf: func(format string, v ...interface{}) {
-			// 	fmt.Printf(format, v)
-			// },
-			// TLS: &tls.Config{
-			// 	InsecureSkipVerify: true,
-			// },
-		})
-	)
-
-	if err != nil {
-		return nil, err
-	}
-
-	if err := conn.Ping(ctx); err != nil {
-		if exception, ok := err.(*clickhouse.Exception); ok {
-			fmt.Printf("Exception [%d] %s \n%s\n", exception.Code, exception.Message, exception.StackTrace)
-		}
-		return nil, err
-	}
-	return conn, nil
-}
-
 func (im *Importer) consumeMessages(pid int) {
 	defer im.wg.Done()
 	defer log.Println("Stopping worker", pid)
 	log.Println("Starting worker", pid)
 
-	conn, err := im.connect()
-	if err != nil {
+	defaultServer, ok := im.Config.Clickhouse.Servers["default"]
```

Comment:
Rather than doing this, I'd like to pass an instance of `ClickhouseProvider` into the `Importer`. Then have the default implementation fetch data from the config file.

If we do that, we could create a single `ClickhouseProvider` instance in `main.go` and pass it to any modules that need it. It would also let us avoid another refactor of the importer since we will only be accessing Clickhouse via the interface.

Reply:
@poundifdef  this has similar effect as we are still reading from the config file. The major difference is that `*Importer.Start()` doesn't take the `ClickhouseProvider` parameter because `importer.NewImporter(&C)` already takes the config which includes the servers.

I can modify the `Start()` function to become `*Importer.Start(provider ClickhouseProvider)` and read in `Start(C.Clickhouse)`.

Reply:
Let's do that - I'd suggest adding it as a parameter to `NewImporter()`.